piptools/sync.py

```diff
@@ -156,11 +158,15 @@ def sync(to_install, to_uninstall, verbose=False, dry_run=False, pip_flags=None,
             for ireq in to_install:
                 click.echo("  {}".format(format_requirement(ireq)))
         else:
-            package_args = []
+            # prepare requirement lines
+            req_lines = []
             for ireq in sorted(to_install, key=key_from_ireq):
-                if ireq.editable:
-                    package_args.extend(['-e', str(ireq.link or ireq.req)])
-                else:
-                    package_args.append(str(ireq.req))
-            check_call([pip, 'install'] + pip_flags + install_flags + package_args)
+                req_lines.append(make_line_from_ireq(ireq))
+
+            # save requirement lines to a temporary file
+            tmp_req_file = tempfile.NamedTemporaryFile(mode='wt')
```

Comment:
You can use `NamedTemporaryFile` as a contenxt manager to ensure it is always deleted when finished:

```py
with tempfile.NamedTemporaryFile('w') as tmp_req_file:
    tmp_req_file.write('\n'.join(req_lines))
    tmp_req_file.flush()

    check_call([pip, 'install', '-r', tmp_req_file.name] + pip_flags + install_flags)
```